src/InMemoryMessageRepository.php

```diff
@@ -70,4 +72,16 @@ public function retrieveAllAfterVersion(AggregateRootId $id, int $aggregateRootV
 
         return $lastMessage instanceof Message ? $lastMessage->aggregateVersion() : 0;
     }
+
+    public function paginate(int $perPage, PaginationCursor | OffsetCursor $cursor): Generator
+    {
+        if( ! $cursor instanceof OffsetCursor){
+            throw new \InvalidArgumentException('Cursor must be an instance of OffsetCursor');
+        }
+
+        $page = array_slice($this->messages, $cursor->offset(), $perPage, false);
```

Comment:
Along with my comment below about `$perPage`, this would become:

```suggestion
        $page = array_slice($this->messages, $cursor->offset(), $cursor->perPage(), false);
```

Reply:
this was implemented as limit()